text/0000-match-ergonmics.md

```diff
@@ -0,0 +1,531 @@
+
+The target expression (aka discriminant) in a match expression is `expr` in
+`match expr { ... }`.
+
+If the target expression is borrowed data, then the compiler will implicitly
+dereference the target expression as many times as necessary to make every
+pattern in every arm type check. Dereference here is equivalent to using the `*`
+operator: either a primitive pointer dereference or calling the `deref` method,
+if a type implements the `Deref` trait.
+
+Auto-deref allows the programmer to access borrowed data within a limited scope
+without having to explicitly dereference the data. This is analogous to the
+implicit dereferencing performed by the dot operator, however, the scope of
+dereference is inline, rather than defined by a function.
+
+More formally, let `T0...Tn` be the types of the patterns of each arm in the
+match expression. If the pattern is a wildcard or variable, we ignore its type.
+Let `expr` be the target expression.
+
+I will use `A <: B` to mean `A` is a subtype of `B`, and `e: T` to mean that
+expression `e` has type `T`. I use `*^m` to mean applying the `*` operator `m`
+times where `m >= 0`.
+
+```
+forall i in [0..n] Ti <: U
+*^m expr: U
+...
+--------------------------
+match expr { ... }: ...
+```
+
+If this rule can be satisfied with different values of `m` (e.g., if none of the
+arms constrain the type), then the compiler will choose the smallest value.
+
+Note that this dereferencing allows the compiler to insert calls to the `deref`
+method (if there is a `Deref` implementation) for any intermediate type found
+whilst dereferencing.
+
+We do not permit auto-referencing. Since `deref` is implemented for `&self`, to
+apply the auto-deref, `expr` must be a borrowed reference. E.g., if `expr` has
+type `Rc<Foo>` we will not implicitly dereference to `Foo`, but we will if the
+type is `&Rc<Foo>`. This means that we never implicitly borrow owned data, which
+should make this conversion easier to reason about. It also matches the
+behaviour of deref coercions (although it is weaker than the dot operator
+conversions).
+
+A less conservative alternative would be to allow auto-referencing. I believe
+there are no technical obstacles. That would be more flexible and match the
+behaviour of the dot operator. However, I believe constraining to borrowed data
+makes reasoning easier. We can always adopt auto-referencing later.
+
+The above implies that the patterns in each arm must have the same type. For
+example, the following match would not type check:
+
+```
+match x: &Option<_> {
+    Some(_) => {}
+    &None => {}
+}
+```
+
+
+### Type inference and backwards compatibility
+
+Currently, we might infer the type of the target expression from the type of the
+patterns. In order to be backwards compatible, we must be able to fall back to
+the type known from the patterns if type inference is stuck without that
+information. I think this will only require choosing `m = 0` if type inference
+can't make progress.
```

Comment:
The case I'm covering here is if we get stuck (if we can proceed by forcing inference, we should) and can't find `m`/`k`. Then we arbitrarily choose 0, and try to continue.